### PR TITLE
Event for trade volume calculation

### PR DESCRIPTION
**Pull request by josojo**
Anxo noted in his review:
Listing trades: This queries would be basic usage of the exchange

And we discussed later on that we want to emit Events such that we can calculate the trading volume of each order.
I created this event, which contains the  batch index and the objective value. Hence, we can filter for the settlement events with the highest objective value per batchIndex and then easily calculate the trading volumes used for settling a batch

### CODE REVIEW

**Review by fleupold — COMMENTED** [review]
I think the plan was to create an event that can be indexed by user and orderId so that it can be easily queried. Can we emit a separate event for each execution and reversion of a trade?

cc @anxolin

**Comment by bh2smith** [comment]
Is this related to #387 ?

**Review by anxolin — CHANGES_REQUESTED** [review]
Yes, exactly as Felix said.
Ideally, the point is being able to get the partial executions for a given order.  So indexed by owner and order Id. Also by batchId to be able to get any Trade by it's id

File: contracts/BatchExchange.sol [hunk 1]
```diff
@@ -108,6 +108,14 @@ contract BatchExchange is EpochTokenLocker {
      */
     event OrderDeletion(address owner, uint256 id);
 
+    /** @dev Event emitted when a new trade is settled
+     */
+    event Trade(address indexed owner, uint256 indexed orderIds, uint256 executedSellAmount, uint256 executedBuyAmount);
+
+    /** @dev Event emitted when an already exectued trade gets reverted
+     */
+    event TradeRevertion(address indexed owner, uint256 indexed orderIds, uint256 executedSellAmount, uint256 executedBuyAmount);
```
Comment (fleupold) [inline contracts/BatchExchange.sol]: nit: naming TradeRever**s**ion

**Review by fleupold — APPROVED** [review]

**Comment by anxolin** [comment]
This one is my favourite of all the changes :) 
thanks

**Comment by anxolin** [comment]
@josojo @fleupold I see you didn't add the batchId in the event. Was this a conscious decission?

**Comment by josojo** [comment]
Yeah, it would be a lot of data duplication. 

From the time of the event, we can later easily determine the batchIndex. At least in dune, it was always very easy.
If you need it for the front-end let us know, then we will change it later

**Comment by anxolin** [comment]
But if it is deterministic, why is in the `submitSolution`

Is the time of the event telling us deterministically the batchId?

**Comment by fleupold** [comment]
> But if it is deterministic, why is in the submitSolution

In case you submitted a solution with a low gas price, it might be stuck in the tx queue for a while and only get executed once the batch it was valid for is already solved. 

In order to fail early (and not use more gas than necessary for a benign race condition) we ask the solver to specify which batch they were intending to solve and can fail early if this batch is no longer accepting solutions.

**Comment by anxolin** [comment]
Okay, so you are saying, it's really 100% deterministic, this id is just as a fail early mechanism and we don't need to explicitly add it in the event.

If it's like that, it's fine for me, we can abstract this somehow in the libraries or when we index the event in the graph

**Pull request merged**